_episodes/l2_02_basic_unit_testing.md

```diff
@@ -0,0 +1,285 @@
+
+4. **Test Data Organization**: Separate the test data from the test methods. Consider using dedicated variables or data structures to store test data. This allows for better readability and maintainability, as changes to the test data can be easily managed without modifying the test methods themselves.
+
+## 7. Unit Test for the Fibonacci Sequence
+
+With the above guidelines in mind, we can write some of the unit tests for our `fibonacci.cpp` that we learnt in Chapter 1. The complete code is given below and can be found in [Chapter2](../code/Chapter2/Fibonacci_test.cpp).
+
+```cpp
+#include <iostream>
+#include "gtest/gtest.h"
+
+/*Returns the n'th term of the Fibonacci sequence.*/
+int recursive_fibonacci(int n)
+{
+    if (n <= 1) {
+        return n;
+    } else {
+        return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2);
+    }
+}
+
+TEST(FibonacciTest, HandlesZeroInput) {
+    EXPECT_EQ(recursive_fibonacci(0), 0);
+}
+
+TEST(FibonacciTest, HandlesValueOneAsInput) {
+    EXPECT_EQ(recursive_fibonacci(1), 1);
+}
+
+TEST(FibonacciTest, HandlesPositiveInput) {
+    EXPECT_EQ(recursive_fibonacci(5), 5);
+}
+
+int main(int argc, char **argv)
+{
+    testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
+```
+
+On compiling and running above program, we get the following output.
+
+```bash
+[==========] Running 3 tests from 1 test suite.
+[----------] Global test environment set-up.
+[----------] 3 tests from FibonacciTest
+[ RUN      ] FibonacciTest.HandlesZeroInput
+[       OK ] FibonacciTest.HandlesZeroInput (0 ms)
+[ RUN      ] FibonacciTest.HandlesValueOneAsInput
+[       OK ] FibonacciTest.HandlesValueOneAsInput (0 ms)
+[ RUN      ] FibonacciTest.HandlesPositiveInput
+[       OK ] FibonacciTest.HandlesPositiveInput (0 ms)
+[----------] 3 tests from FibonacciTest (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 3 tests from 1 test suite ran. (0 ms total)
+[  PASSED  ] 3 tests.
+```
+### 7.1 Exercise
+
+Modify the above program to throw an exception for negative values and write a test for this.
+
+**Solution**
+
+Add the following lines in your code. The full solution is given in [Solution](../code/Chapter2/Exercise_solution).
+
+```cpp
+// Change the function as shown below.
+int recursive_fibonacci(int n)
+{
+    if(n < 0)
+        throw std::invalid_argument("Input must be a non-negative number");
+    else if (n <= 1) {
+        return n;
+    } else {
+        return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2);
+    }
+}
+
+// Add the required test.
+TEST(FibonacciTest, ThrowsExceptionNegativeInput) {
+    EXPECT_THROW(recursive_fibonacci(-3), std::invalid_argument);
+}
+```
```

Comment:
This needs to be formatted as a challenge with a solution inside, so it is rendered correctly in GitHub pages:

```suggestion
> ## Exercise: Test for negative values
> 
> Modify the above program to throw an exception for negative values and write a test for this.
>
> > ## Solution
> > 
> > Add the following lines in your code. The full solution is given in [Solution](../code/Chapter2/Exercise_solution).
> > 
> > ```cpp
> > // Change the function as shown below.
> > int recursive_fibonacci(int n)
> > {
> >     if(n < 0)
> >         throw std::invalid_argument("Input must be a non-negative number");
> >     else if (n <= 1) {
> >         return n;
> >     } else {
> >         return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2);
> >     }
> > }
> > 
> > // Add the required test.
> > TEST(FibonacciTest, ThrowsExceptionNegativeInput) {
> >     EXPECT_THROW(recursive_fibonacci(-3), std::invalid_argument);
> > }
> > ```
> >
> {: .solution}
>
{: .challenge}
```